src/include/sof/audio/audio_stream.h

```diff
@@ -200,6 +200,9 @@ static inline void audio_stream_reset(struct audio_stream *buffer)
 static inline void audio_stream_init(struct audio_stream *buffer,
 				     void *buff_addr, uint32_t size)
 {
+	assert(buffer);
+	assert(buff_addr);
+
```

Comment:
I don't think this is a good idea. This means that all the functions that get pointers as arguments need to check for NULLs. Better idea would be to check for correct memory allocation immediately after the allocation.

Even, with this assert the current testing code [test/cmocka/src/audio/selector/selector_test.c ] will crash:

»       /* allocate new sink buffer */                                                                                                                                                                             
»       sel_state->sink = test_malloc(sizeof(*sel_state->sink));
»       sel_state->sink->frame_fmt = parameters->sink_format;
»       pbuff = test_calloc(parameters->buffer_size_ms, size);
»       audio_stream_init(sel_state->sink, pbuff,
»       »       »         parameters->buffer_size_ms * size);

```

Reply:
@dbaluta UT tests on this branch pass and I can't see your point. 

> This means that all the functions that get pointers as arguments need to check for NULLs
In my opinion it's brilliant idea - this indicate professional code style.

> Better idea would be to check for correct memory allocation immediately after the allocation.
And what about for example wrong pointers handling or passing variable value instead of address?

>Notice that sel_state->sink is accessed long before we actually check for it.
In this place of code yes, in another maybe no

Reply:
@ktrzcinx my expectation is that any user configurable resource allocation failure should **never** assert or panic but should be returned up the stack (to the user) as an error. This means any topology, stream configuration or mixer data i.e. any data the user inputs at runtime.
I do expect panics and asserts for failures of static core build time configuration resources that are critical to overall FW usage. i.e. buffer for IRQ core, buffer for trace core, IRQ allocation for scheduler, etc

Reply:
@lgirdwood Sure, but lets see that during allocation should be resolved in component where allocation is called (in this example comp_buffer as usual) and then returned to up the stack. In this place allocation of memory doesn't happen, so if here somebody passed null pointer, it can be return up the stack with introduced api. So or we will call panic here or write on random memory regions and spend time for debugging.

Reply:
Checks can be done better than with assertions. I think in the future assertions can be disabled (they certainly can if we move to Zephyr), as in error return/cleanup paths just like in the kernel. So we cannot rely on asserts anyway.

Reply:
I'm not willing to remove allocation error from other places and base on this two assertions, it is a way to rapidly find error when there will be some bug in implementations on higher layers of api - last line of defence. Everything should work as well when assertions will be disabled in release configuration.

Reply:
The thing is, some asserts can end up being redundant. I'm not sure so I do accept these but some assertions can be redundant, and at least on my platform I do NOT see stack traces for panics so if an assert triggers for me I'm unable to see it. But if there is a higher level assert which would also catch the same wrong thing there is no need for the lower level one. Having asserts everywhere may not be the best thing, I for one inserted none in my driver code because everything that needs to be checked and asserted is asserted by the higher level code.

So I'm not sure of the big picture, whether this buffer is also verified somewhere else before reaching this (and thus these asserts are worth it or not), that's why I approved.

Reply:
So if it's ever possible for something you assert to be non-NULL to become NULL for any reason other than bugs, then you should put a check rather than an assert.

Reply:
> @ktrzcinx my expectation is that any user configurable resource allocation failure should **never** assert or panic but should be returned up the stack (to the user) as an error. This means any topology, stream configuration or mixer data i.e. any data the user inputs at runtime.
> I do expect panics and asserts for failures of static core build time configuration resources that are critical to overall FW usage. i.e. buffer for IRQ core, buffer for trace core, IRQ allocation for scheduler, etc

Most of components have asserts in flows that are user-initiated, so your expectation is already unsatisfied. I don't want to argue if they are good or bad, but in kernel we have no asserts at all in our driver, so yea maybe it's better to limit them to the cases where static-routine failed and for other flows always bubble up error.

Reply:
@jajanusz and we need to fix this, I had a feeling it was already not being followed in some places ! We should probably think about it for v1.6
@ktrzcinx I understand that the assert can be handy for debug, but it's fatality is not desired for production, this will be addressed in future trace updates where we can vary the level and content at runtime.

Reply:
Ok, I close this PR